KHLY2700/KHLY2700-IOC-01App/Db/keithley2700.db

```diff
@@ -295,151 +289,173 @@ record(bi, "$(P)BUFF:AUTOCLEAR")
-	info(INTEREST, "LOW")
-	field(EGU, "")
+record(calcout,  "$(P)INDEX:START:CALC") {
+  field(DESC,  "Where to start reading from")
+  field(INPA, "$(P)BUFF:NEXT")
+  field(INPB, "$(P)INDEX:START")
+  # If BUFF:NEXT is less than INDEX:START AND BUFF:NEXT is greater than 0
+  # then the buffer has cleared and more readings have been added, so
+  # begin reading from 0. Otherwise, read as normal
+  field(CALC,  "((A<B)&&(A>0))?0:B")
+  field(OUT,  "$(P)INDEX:START PP")
 }
 
-record(ao, "$(P)CURR:LOC")
-{
-	field(DESC, "Current buffer read location")
-	info(INTEREST, "LOW")
-	field(EGU, "")
+record(ao, "$(P)INDEX:START") {
+  field(DESC, "Index loc to start retrieve readings")
+  field(PINI,  "YES")
+  field(VAL, "0")
+  field(FLNK, "$(P)COUNT:CALC")
 }
 
-record(calcout, "$(P)COUNT:CALC") 
-{
-	field(DESC, "Calculating number of readings")
-	field(OOPT, "Every Time")
-	field(INPA, "$(P)BUFF:NEXT")
-	field(INPB, "$(P)CURR:LOC")
-	field(CALC, "A-B")
-	field(OUT, "$(P)COUNT PP")	
-	field(FLNK, "$(P)BUFF:READ")
+record(calcout, "$(P)COUNT:CALC") {
+  field(DESC, "calculates number of readings to return")
+  field(INPA, "$(P)BUFF:NEXT")
+  field(INPB, "$(P)INDEX:START")
+  field(INPC, "$(P)BUFF:SIZE")
+  # if next free buff location is >= index to start reading from, then this is
+  # the normal case, and count is simply next free buffer location - starting
+  # index. If not, then check if next buffer location is reported as 0. If it is
+  # then buffer has filled and there are readings yet to be retrieved from the
+  # buffer, so retrieve them. If next buffer location is not 0, then readings
+  # have been cleared from the buffer, so just read from index 0 up to next free
+  # buffer location. See Wiki page for more detail
+  field(CALC, "(A>=B)?(A-B):((A=0)?(C-B):A)")
+  field(OUT, "$(P)COUNT PP")
 }
 
-record(waveform, "$(P)BUFF:READ")
-{
-    field(DESC, "Buffer readings from start to +count")
-	field(NELM, "100")
-	field(FTVL, "DOUBLE")
-    field(DTYP, "stream")
-    field(INP,  "@devKeithley_2700.proto getBuffReadings($(P)) $(PORT)")
-	info(ARCHIVE, "VAL")
-    info(INTEREST, "HIGH")
-	field(PREC, "5")
-	field(FLNK, "$(P)CURR:POS:CALC")
-    field(SIML, "$(P)SIM")
-    field(SIOL, "$(P)SIM:BUFF:READ")
-    field(SDIS, "$(P)DISABLE")
+record(ao, "$(P)COUNT") {
+  field(DESC, "number of items to return")
+  field(FLNK,  "$(P)BUFF:READ:TRIG")
 }
 
-record(calcout, "$(P)CURR:POS:CALC")
-{
-	field(DESC, "Update current buffer memory location")
-	field(INPA, "$(P)BUFF:NEXT")
-	field(CALC, "A")
-	field(OUT, "$(P)CURR:LOC PP")
-	field(OOPT, "Every Time")
-	field(FLNK, "$(P)BUFF:READ:PARSE:READINGS")
+record(calcout, "$(P)BUFF:READ:TRIG") {
+  field(INPA,  "$(P)COUNT")
+  field(CALC,  "A>0?1:0")
+  # forward link predicated on COUNT>0 ?
+  field(OOPT,  "When Non-zero")
+  field(OUT,  "$(P)BUFF:READ PP") # this will send 1 to BUFF:READ
 }
 
-# These aSub records parse BUFF:READ into 
+record(waveform, "$(P)BUFF:READ")
```

Comment:
Add a comment that clarifies the format of the data.

Reply:
Added